### PR TITLE
podUnavailableBudget apis and validating

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: liheng.zms <liheng.zms@alibaba-inc.com>

### Ⅰ. Describe what this PR does
podUnavailableBudget apis and validating

### CODE REVIEW

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	allErrs := h.validatingSidecarSetFn(obj)
+	if len(allErrs) != 0 {
+		return admission.Errored(http.StatusInternalServerError, allErrs.ToAggregate())
+	}
+	return admission.ValidationResponse(true, "")
```
Comment (FillZpp) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:62]: should we also validate update operation that forbid users to modify `spec.selector` or `spec. targetRef `?

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		allErrs = append(allErrs, validatePubConflict(obj, pubList.Items, field.NewPath("spec"))...)
+	}
+	return allErrs
+}
+
+func validatePodUnavailableBudgetSpec(obj *policyv1alpha1.PodUnavailableBudget, fldPath *field.Path) field.ErrorList {
+	spec := &obj.Spec
+	allErrs := field.ErrorList{}
+
+	if spec.Selector == nil && spec.TargetReference == nil {
```
Comment (FillZpp) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:83]: should also validate `spec.Selector != nil && spec.TargetReference != nil` for they are mutually exclusive

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("TargetReference"), spec.TargetReference, "empty TargetReference is not valid for PodUnavailableBudget."))
+		}
+	} else {
+		allErrs = append(allErrs, metavalidation.ValidateLabelSelector(spec.Selector, fldPath.Child("selector"))...)
+		if len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is not valid for PodUnavailableBudget."))
+		}
+	}
+
+	if spec.MaxUnavailable == nil && spec.MinAvailable == nil {
```
Comment (FillZpp) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:96]: should also validate `spec.MaxUnavailable != nil && spec.MinAvailable != nil` for they are mutually exclusive

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		old = new(policyv1alpha1.PodUnavailableBudget)
+		if err := h.Decoder.Decode(
+			admission.Request{AdmissionRequest: admissionv1beta1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}},
+			old); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+	allErrs := h.validatingSidecarSetFn(obj, old)
+	if len(allErrs) != 0 {
+		return admission.Errored(http.StatusInternalServerError, allErrs.ToAggregate())
```
Comment (furykerry) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:71]: is it better to use. http.StatusBadRequest for some error cases here ?

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	}
+	return admission.ValidationResponse(true, "")
+}
+
+func (h *PodUnavailableBudgetCreateUpdateHandler) validatingSidecarSetFn(obj, old *policyv1alpha1.PodUnavailableBudget) field.ErrorList {
```
Comment (furykerry) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:76]: please change the function name

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		allErrs = append(allErrs, field.InternalError(field.NewPath(""), fmt.Errorf("query other podUnavailableBudget failed, err: %v", err)))
+	} else {
+		allErrs = append(allErrs, validatePubConflict(obj, pubList.Items, field.NewPath("spec"))...)
+	}
+	return allErrs
+}
+
+func validateUpdatePubConflict(obj, old *policyv1alpha1.PodUnavailableBudget, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !reflect.DeepEqual(obj.Spec.Selector, old.Spec.Selector) || !reflect.DeepEqual(obj.Spec.TargetReference, old.Spec.TargetReference) {
```
Comment (furykerry) [inline pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:95]: shall we check whether the pod scope specified by two selector has some overlapping cases?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go (hunk 6 above)
Comment (zmberg) [reply pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go:95]: Is there a scenario where the system owner has created a high-level (project-level) pub protection policy and the owner of a specific module has created a more fine-grained pub, but in fact both pubs do not conflict and can work simultaneously?

**Review by furykerry — APPROVED** [review]

**Pull request merged**